### PR TITLE
Mapping to segment container tracks

### PR DESCRIPTION
**Pull request by philipnbbc**
# Details
This PR adds an ADR that proposes adding a `container_mapping` property to the core flow metadata that provides additional metadata to help identify which track (and component in a track) provides essence for the Flow. The proposal is that a number of mapping types can be used together to support readers with different capabilities.

Note that the assumption is that the mapping is for the container that holds the essence, not the container that provided the essence and was copied to the store for the Flow. I.e. the mapping to the tracks of the ingested stream will be lost without a relationship between the Flow for the stored content and the Flow associated with the ingested stream.

The assumption is that details of the timing mapping are held elsewhere. E.g. the Flow Segment has the `ts_offset` that is used to map the timing from MPEG-TS segment containers. E.g. tags are currently used to indicate that TEMI timeline descriptor are embedded in MPEG-TS.

# Pivotal Story
Story URL: https://www.pivotaltracker.com/story/show/187323816

# Related PRs
_Where appropriate. Indicate order to be merged._

# Submitter PR Checks
_(tick as appropriate)_

- [ ] PR completes task/fixes bug
- [ ] API version has been incremented if necessary
- [ ] ADR status has been updated, and ADR implementation has been recorded
- [ ] Documentation updated (README, etc.)
- [ ] PR added to Pivotal story (if relevant)
- [ ] Follow-up stories added to Pivotal

# Reviewer PR Checks
_(tick as appropriate)_

- [ ] PR completes task/fixes bug
- [ ] Design makes sense, and fits with our current code base
- [ ] Code is easy to follow
- [ ] PR size is sensible
- [ ] Commit history is sensible and tidy

# Info on PRs
The checks above are guidelines. They don't all have to be ticked, but they should all have been considered.

### CODE REVIEW

**Comment by samdbmg** [comment]
As discussed, this makes a lot of sense as it is. Could you write some Flow examples that use the two flavours of this? Feel free to modify the existing examples, but it would be nice to make an internally consistent set of Flows

**Comment by philipnbbc** [comment]
I've added some examples which are referenced from the spec. `/flows/<flow ID>` endpoint. An application note is also included the describes the use of `container` and `container_mapping` in the context of multi-essence Flows.

File: docs/appnotes/0002-containers-and-mappings.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+The presence of the `container` property indicates that the Flow has Flow Segments (the list could be empty).
+
+![Diagram showing a Flow referencing media objects containing a video track only](./images/0002-containers-and-mappings-fig1.png)
+
+The diagram below shows a sequence of MPEG-TS media objects that contains a multplex of audio and video.
+The multi-essence Flow references the media objects via a sequence of Flow Segments.
+The multi-essence Flow has a `container` property that indicates that this Flow has Flow Segments.
+The video and audio Flows are referenced by a Collection Item in the `flow_collection` list.
+The video and audio Flows don't themselves have a sequence of Flow Segments and therefore the `container` property is not set.
+The media content for these Flows are accessed using the multi-essence Flow's Flow Segments.
```
Comment (samdbmg) [inline docs/appnotes/0002-containers-and-mappings.md:32]: Nit: Does something like this make this paragraph less dense?
```suggestion
The diagram below shows a sequence of MPEG-TS media objects that contains a multplex of audio and video.
The multi-essence Flow:
  - has a `container` property that indicates that this Flow has Flow Segments.
  - uses its sequence of Flow Segments to reference the media objects.

The video and audio Flows:
  - are referenced by a Collection Item in the `flow_collection` list.
  - don't themselves have a sequence of Flow Segments and therefore the `container` property is not set.
The media content for these Flows are accessed using the multi-essence Flow's Flow Segments.
```

File: docs/appnotes/0002-containers-and-mappings.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
+# 0002: Containers and Mappings
+
+## Abstract
+
+Flows reference media content held in media objects in TAMS.
+The media objects containers may wrap media content from multiple Flows.
```
[the remark this answers is not in the document]
Comment (philipnbbc) [reply docs/appnotes/0002-containers-and-mappings.md]: I changed the wording a bit more to make it clearer hopefully: https://github.com/bbc/tams/pull/46/commits/eb085a5e8cf8a3d4c1ccaed60cd17c4fb9bc8e66

**Comment by philipnbbc** [comment]
> On the versioning, I'm not sure this is a breaking change? Making `container` mandatory would be, but making optional something that was previously required (while introducing a meaning to it not being set) feels like adding a new feature, rather than breaking an assumption?

I was thinking that from the reading side of the API that it would a breaking change because readers would expect the property to be present. As you say, from the writing side it isn't a breaking changed. I think if the property was new and optional it would be a new feature.

**Comment by philipnbbc** [comment]
> Since we're unlikely to implement this into Squirrel before next week, should this one be left open for now, and it'll form a new TAMS version after v4.0

Agree, I'll leave it open.

**Pull request merged**